addons/csw/functions/fnc_autofire_fire.sqf

```diff
@@ -0,0 +1,109 @@
+#include "..\script_component.hpp"
+/*
+ * Author: tcvm
+ * Fires the weapon.
+ *
+ * Arguments:
+ * 0: CSW <OBJECT>
+ * 1: Muzzle to shoot <STRING>
+ *
+ * Return Value:
+ * None
+ *
+ * Example:
+ * [csw, "Main"] call ace_csw_fnc_autofire_fire
+ *
+ * Public: No
+ */
+params ["_csw", "_weapon", "_muzzle"];
+
+TRACE_5("autofire_fire",_csw,_weapon,_muzzle,gunner _csw,local gunner _csw);
+
+private _gunner = gunner _csw;
+if (!local _gunner || !local _csw) exitWith {};
+
+private _fireMode = currentWeaponMode _gunner;
+if (_fireMode isEqualTo "") then {
+    private _cswHashmap = _csw getVariable [QGVAR(autofire_defaultModes), createHashMap];
+    private _needsPropogate = false;
+    _fireMode = _cswHashmap getOrDefaultCall [
+        _muzzle,
+        {
+            _needsPropogate = true;
+            private _defaultModes = getArray (configFile >> "CfgWeapons" >> _muzzle >> "modes");
+            TRACE_3("setting default mode",_csw,_muzzle,_defaultModes);
+            if (_defaultModes isEqualTo []) exitWith { "" };
+            _defaultModes select 0
+        },
+        true
+    ];
+
+    if (_needsPropogate) then {
+        _csw setVariable [QGVAR(autofire_defaultModes), _cswHashmap, true];
+    };
+};
+
+_gunner setVariable [QGVAR(autofire_isProxy), true];
+
+// we have to adjust fire if we aren't a player shooting
+// because the engine does some fuckery with players, AI shoot down the barrel, but players shoot with some offset
+private _fireEventHandler = _csw addEventHandler [
+    "Fired",
+    {
+        params ["_csw", "_weapon", "", "", "", "", "_projectile", "_gunner"];
+
+        // disable for non-proxy and players, this is already done for us
+        if ([_gunner] call EFUNC(common,isPlayer) || { !(_gunner getVariable [QGVAR(isProxy), false ])}) exitWith {};
+
+        (_csw getVariable [QGVAR(autofire_animations), ["", ""]]) params ["_body", "_gun"];
+
+        private _weaponDir = _csw weaponDirection _weapon;
+
+        private _traverse = _csw animationSourcePhase _body;
+        private _turretRot = [vectorDir _csw, vectorUp _csw, deg _traverse] call CBA_fnc_vectRotate3D;
+        private _elevation = (acos ((_turretRot vectorCos _weaponDir) min 1)) + ((_turretRot call CBA_fnc_vect2polar) select 2);
+
+        private _elev = _csw animationSourcePhase _gun;
+
+        private _lookDirection = _csw vectorModelToWorldVisual ([1, -deg _traverse, deg _elev] call CBA_fnc_polar2vect);
+        private _azimuth = (_lookDirection select 0) atan2 (_lookDirection select 1);
+
+        private _projectileSpeed = vectorMagnitude velocity _projectile;
+        private _barrelDirection = [1, _azimuth, _elevation] call CBA_fnc_polar2vect;
+
```

Comment:
not sure, do we need to worry about re-adding random dispersion?

Reply:
yes we do, i added it now but idk if it's how arma simulates it. I find shells will still be dispersed, but VERY shallow-y. 0.04 degree offset sort of deal